### ISSUE TITLE
CPU usage keeps increasing

### ISSUE THREAD

**Issue by pjchavarria** [ParsePlatform/parse-server#1878]
#### Environment Setup

- Server: parse-server version, operating system, hardware, local or remote?
Parse Server running on AWS Elastic Beanstalk (Amazon Linux) using EC2 t2.medium.
    "parse": "~1.8.1",
    "parse-dashboard": "^1.0.11",
    "parse-server": "~2.2.6"

- Database: version, storage engine, hardware, local or remote?
Database 3.0.11 running on mLab.com

#### Steps to reproduce
- Put the app on production.
- The CPU usage will start increasing. After restarting the instance the CPU usage drops to the initial level ~10%.

I've read it could be cloud code not handling all responses, but i've also tried **disabling all cloud code on production**, getting the same result.

In the following image you can see the 4 spikes of CPU usage, the drop happens when i redeploy the application.
As you can see this is not related to the number of requests nor data in/out.
![image](https://cloud.githubusercontent.com/assets/1723531/15456181/672e9f14-201d-11e6-904f-17a8f3057e25.png)
The drop of CPU usage in the image is after restarting the server, similar to #1476

**Comment by iForests** [comment]
I have similar problem, and I solve it by increasing the instance number running on AWS.

**Comment by huang1196** [comment]
Same problem with u. Still waiting for the official response. I have to restart my server every day.

**Comment by themichael123** [comment]
Hey Guys. This is the garbage collection kicking in due to the many post requests fired off by parse server. Every time a post request is being fired of this creates a new instance of the entire application. Increasing CPU usage will hence go hand in hand with increasing memory usage (not shown in the dashboard) and increasing latency. If you run this in production and do not want to rewrite your application using a nodejs mongo/redis driver, a possible solution (that we use) is to install PM2 to essentially load balance on a per server level and restart the node app after reaching 100M of memory consumption. It is crucial to do this with multiple cores as the PM2 load balancer will handle the restart gracefully (nodejs is single threaded/runs on one core).

**Comment by airdrummingfool** [comment]
When I was running only 2 instances, I experienced this issue often. I never saw RAM usage runaway increase (even though CPU usage & latency did greatly increase). After bumping up to 3 instances, the CPU/latency issues stopped. I haven't restarted in almost 2 weeks, and serve approximately 1-2k requests per minute, and everything seems good.

@themichael123 if that is true, is there an easy test case to reproduce this behavior? I wasn't able to recreate it on an EC2 instance by hammering one route (a query for the first object of one class) with 4 million requests.

**Comment by pjchavarria** [comment]
Not sure if this is the issue, but reading the logs i get this error many times:
```/var/log/nodejs/nodejs.log
-----
ESC[31merrorESC[39m: error while sending push code=101, message=Object not found.
ESC[31merrorESC[39m: error while sending push code=101, message=Object not found.
.
.
```

I've made sure the to handle all cases where an exception could occur on cloud code (using promises). Also exiting early if any request param is missing.
```
if (!postedBy || !toUserID  || !pushMessage || !targetID) {
        response.error("sendNewTimelinePostReplyPushNotificationV2 missing params");
        return
}```

**Comment by themichael123** [comment]
1. Host an express route at app.get /something. 
2. Complete some tasks, query the db etc. and call res.render to serve a webpage.

 This will lead to memory issues and the garbage collection kicking in. As we serve millions of pageviews every day the garbage collection will fall behind and get slower and slower. Parse server is not particularly efficient for this use case and we will have to migrate to native mongo and redis + express.

**Comment by steven-supersolid** [comment]
We see increased memory usage over time, thought to be due to caching of _User as usage flattens off after around a day. I'd expect a memory leak to have a steady increase on top of that. CPU usage and latency approximately doubles over the same period. We're on Heroku so after 24 hours the instances cycle.

I'm not fully familiar with garbage collection in Node.js yet but would expect memory usage to drop when the GC kicks in and then steadily rise again. I'd expect a slight increase CPU usage while performing the GC. If it's possible to output some debug for GC then I could be certain that is the issue for my setup.

It seems like we all have different setups and each parse system we use is a potential source of this issue. Focusing on CPU usage seems to be the way to go.

**Comment by iForests** [comment]
Here is what I saw when the CPU usage was at the peak:

- Latency was very high (59 sec).

- In `/var/log/nginx/error.log`, a lot of `1024 worker_connections are not enough`:
```
2016/05/27 04:03:17 [alert] 2717#0: 1024 worker_connections are not enough
2016/05/27 04:03:17 [alert] 2718#0: 1024 worker_connections are not enough
```
- In `/var/log/nginx/access.log`, a lot of requests failed with error code 499:
```
10.233.155.30 - - [27/May/2016:04:03:17 +0000] "POST /parse/classes/_Installation/ZgQyjSAwjn HTTP/1.1" 499 0 "-" "node-XMLHttpRequest, Parse/js1.8.5 (NodeJS 4.4.3)" "54.146.46.233"
10.233.155.30 - - [27/May/2016:04:03:17 +0000] "POST /parse/functions/getInbox HTTP/1.1" 499 0 "-" "Parse Android SDK 1.13.0 (app.xxx/124) API Level 23" "24.165.125.220"
```
- In `/var/log/nodejs/nodejs.log`, a lot of `502 Bad Gateway` returned by Nginx:
```
[ERROR][@NF](newsfeed.js:1014) in saveNFR : [null,{"code":107,"message":"Received an error with invalid JSON from Parse: <html>\r\n<head><title>502 Bad Gateway</title></head>\r\n<body bgcolor=\"white\">\r\n<center><h1>502 Bad Gateway</h1></center>\r\n<hr><center>nginx/1.8.1</center>\r\n</body>\r\n</html>\r\n","giveUp":true}]
[ERROR][@getPost](getPost.js:242) Received an error with invalid JSON from Parse: <html><head><title>502 Bad Gateway</title></head><body bgcolor="white"><center><h1>502 Bad Gateway</h1></center><hr><center>nginx/1.8.1</center></body></html>
```

And adding a lot of new instances fixed the problem (it doesn't work if you only add one or two new instances, maybe since there're still too many requests needed to be processed).

**Comment by flovilmart** [comment]
Closing due to lack of activity, please update to latest parse-server version and reopen if the issue persist.

Don't forget to include your current:
- node version
- npm version
- parse-server version
- any relevant logs (VERBOSE=1 will enable verbose logging)

**Comment by mkaya93** [comment]
I have this problem too my parse server version is 2.2.17 and its working with 2 core 7gb ram.Also I'm using pm2 but it's not changed anything.And I'm not getting request yet.My product is under development process. I'm thinking what will happen after I release.

![ekran resmi 2016-07-24 03 20 41](https://cloud.githubusercontent.com/assets/6665723/17080975/b30e4f18-514d-11e6-80ec-733c2a7255f3.png)

**Comment by AKyashkin** [comment]
In our case issue was with multiple calls of Parse.Object.extend("Article") method inside function.
So, basically we have parser, which launches every 5 mins and parses some articles. For each new article we called Parse.Object.extend("Article") inside parse function – what way CPU keept increase util it reached 100% load.

We just moved that call outside the function to global variables and used it as ivar. All problems are gone for now. We run on 1GB RAM DO instance with about 50k users and parser on the same instance in background without any issues.
